doc/design/block.md

```diff
@@ -0,0 +1,128 @@
+
+What's more, Paddle should has a `pd.namespace`, which will help to preventing name conflicts of variables in large model.
+
+For example:
+
+```python
+import paddle as pd
+
+def submodel(namespace, x):
+    # begin a namespace
+    with pd.namespace(namespace):
+        # should be complex network, here easy one for demo
+        # variable W's name is namespace + "W"
+        # so in different namespace, there is a W without conflicts
+        W = pd.get_variable("W", shape=[30, 30], reuse=True)
+        y = pd.matmul(W, x)
+        return y
+
+inputs = [a, b, c]
+
+# create 3 submodels and each one has its own parameters.
+a_out = submodel("model_a", a)
+b_out = submodel("model_b", b)
+c_out = submodel("model_c", c)
+
+
+# d is some op's output
+d = someop()
+
+# reuse model_a
+d_model_a_out = submodel("model_a", d)
+```
+
+with `pd.namespace`, user create 3 different namespaces and in each one, the parameters those have specific name will be reused.
+
+## Why use Block to replace NetOp
+- It is weird to treat Network as an operator, but Block in programming language means grouping multiple operators and acts like a single operator, treat a Block as an operator seems more reasonable.
+- The basic concepts including`Variable`, `Operator` and `Block` are concepts in programming language, but `NetOp` is not.
+- Block is a concept in programming language, it behaves similar to current NetOp, but much concrete, and can guide our futher design.
+  - a block looks like something with a pair of curly bracket `{}`, in python we can use `with` statement to make a dramatic code block.
+  - a block will contains its operators(just like NetOp) and some local variables
+  - operator inside a block can make operation on global variables
+  
+## Implementation
+Currentlly, there is a simple implementation in [User Interface Design](), which has the most function of a Block when user writes a program. 
+But it relays on `NetOp` when executes and that works.
+
+We may rename `NetOp` in cpp code to `Block`, but whether to add more functions of a real block into that implementation depends on whether we want to have a `VarDesc` and split the compilation and execution period.
+
+In my opition, Block is very basic concept in underlying implementation, so a python wrapper and a cpp minimal implementation is enough now.
+It is free to change if other module needs more complex supports such as `VarDesc` is added.
+
+## `with` statement
+Currently Block concept is invisible to users, but some usage are embedded in python interface.
+
+```python
+op = pd.ifelseop()
+
+with op.true_block():
+    # declare inputs
+    a = op.add_input(x)
+    # some complex net
+    b = pd.fc(a)
+    # declare output
+    op.add_output(b)
+    
+with op.false_block():
+    # declare inputs
+    a = op.add_input(x)
+    # some complex net
+    b = pd.fc(a)
+    # declare output
+    op.add_output(b)
+```
+
+```python
+rnn = pd.rnn_op()
```

Comment:
As we analogy the DL system with programming languages in this design, I am thinking that the step net to the RNNOp is like a C++ function `step_net` to a caller function `RNNOp`:

```cpp
void RNNOp(
    const vector<Something>& input, 
    vector<Something>* output, 
    function<Something(Something, int*)> step_net) {
  for (const auto& in_element : input) {
    output->push_back(step_net(in_element, 0/*initial memroy value*/));
  }
}

Something step_net(const Something& input_element) {
  return DoSomething(input_element)
}
```

However, this implementation doesnt' handle the "memory".  In C++, it is true that we can add a memory like:

```cpp
void RNNOp(
    const vector<Something>& input, 
    vector<Something>* output, 
    function<Something(Something, int*)> step_net) {
  Something memory = 0;
  for (const auto& in_element : input) {
    output->push_back(step_net(in_element, 0/*initial memroy value*/));
  }
}

Something step_net(const Something& input_element) {
  memory = Update(input_element, memory);
  return DoSomething(input_element, memory);
}
```

However, above solution doesn't work in a DL system, because each step needs to have its local copy of memory, so could all these memories be kept for being used by the backward algorithm later.  An intuitive solution is to define all these memories in `RNNOp`:

```c++
void RNNOp(
    const vector<Something>& input, 
    vector<Something>* output, 
    function<Something(Something, int*)> step_net) {
  vector<MemoryType> memories;
  memories.resize(input.size()+1);
  memories[0] = kInitialMemoryValue;
  for (int i = 1; i <= input.size(); i++) {
    output->push_back(step_net(input, memories, i));
  }
}

Something step_net(
    const vector<Something>& input, 
    vector<MemoryType>* memories, 
    int step) {
  memories[step] = Update(memories[step-1], input[step]);
  return DoSomething(input[step], memories[step]);
}
```

Some might challenge that `RNNOp` doesn't know `MemoryType` at all -- only `step_net` knows the type.  However, in our design, or, in any dynamic-typed programming languages, both `MemoryType` and `Something` are actually `Variable`.  So it seems that this approach works.

But anyway, it still has a problem -- Block doesn't implement functions, but only the body of a function definition.  To mimic the invocation of `step_net` in `RNNOp`, we need either creates a class `Function` derived from `Block`, or we just don't use function calls in above example -- instead, we can use just C++ blocks:

```c++
void RNNOp(
    const vector<Something>& input, 
    vector<Something>* output) {
  vector<MemoryType> memories;
  memories.push_back(kInitialMemoryValue);
  int step;
  void step_block() { # We use a C++ function without parameter nor return value to mimic block
    memories.push_back(Update(memories[step-1], input[step]));
    output->push_back(DoSomething(input[step], memories[step]));
  }    
  for (int step = 1; step <= input.size(); step++)
    step_block();
}
```

Please be aware that the sub-scope created by `parnet_scope->NewScope()` has a parent pointer and can be used to access all variables defined in `parent_scope`, so our Block mechanism is like defining a void function (a function without parameter nor return value) inside the C++ RNNOp function, so it can access variables defined in RNNOp.

Reply:
Currently, the RNNOp is a rough implementation, and embedded all the details inside operator such as `output->push_back`.

We will have a `RNNStep` operator, which will expose most details as interfaces to python, and maybe in that implementation, some details may be same as this comment.